modules/apps/portal-security/portal-security-content-security-policy/src/main/java/com/liferay/portal/security/content/security/policy/internal/servlet/filter/ContentSecurityPolicyTestFilter.java

```diff
@@ -0,0 +1,242 @@
+/**
+ * SPDX-FileCopyrightText: (c) 2024 Liferay, Inc. https://liferay.com
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+package com.liferay.portal.security.content.security.policy.internal.servlet.filter;
+
+import com.liferay.petra.string.StringBundler;
+import com.liferay.portal.kernel.content.security.policy.ContentSecurityPolicyNonceProviderUtil;
+import com.liferay.portal.kernel.log.Log;
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.portal.kernel.util.ContentTypes;
+import com.liferay.portal.kernel.util.HtmlUtil;
+import com.liferay.portal.kernel.util.PortalRunMode;
+import com.liferay.portal.kernel.util.StringUtil;
+import com.liferay.portal.kernel.util.Validator;
+import com.liferay.portal.security.content.security.policy.internal.ContentSecurityPolicyNonceManager;
+import com.liferay.portal.servlet.filters.BasePortalFilter;
+
+import java.io.PrintWriter;
+
+import java.util.Objects;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Attributes;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * This filters looks for content that has not been adapted to CSP in any HTML
+ * output generated by the Portal.
+ *
+ * It is only enabled when the Portal is started in test mode.
+ *
+ * @author Iván Zaera Avellón
+ * @review
+ */
+@Component(
+	property = {
+		"after-filter=Content Security Policy Filter", "dispatcher=FORWARD",
+		"dispatcher=REQUEST", "servlet-context-name=",
+		"servlet-filter-name=Content Security Policy Test Filter",
+		"url-pattern=/*"
+	},
+	service = Filter.class
+)
+public class ContentSecurityPolicyTestFilter extends BasePortalFilter {
+
+	@Override
+	public boolean isFilterEnabled(
+		HttpServletRequest httpServletRequest,
+		HttpServletResponse httpServletResponse) {
+
+		if (PortalRunMode.isTestMode() ||
+			StringUtil.equals(
+				httpServletRequest.getServerName(), "localhost")) {
```

Comment:
So would this be on whenever developing locally? Would this also be turned on for customers?

Also, should this only be on if CSP is enabled? Or maybe I missed that some place else